### ISSUE TITLE
R.evolve seems to behave badly when the list contains holes

### ISSUE THREAD

**Issue by benzen** [ramda/ramda#1801]
It seems that adjusting will introduce undefined

````
js
const l = []
l[10] = {number: 22}
l[20] = {number: 44}

R.adjust(R.evolve({number: R.inc()}), 20, l)

// => [undefined, undefined, undefined, undefined, undefined, undefined, undefined, undefined, undefined, undefined, {"number": 22}, undefined, undefined, undefined, undefined, undefined, undefined, undefined, undefined, undefined, {"number": 45}]
`````

I understand that it's a default behavior in js.
But there is ways to know which keys in the array are initialized.
I can think of : 

`l.propertyIsEnumerable(102)` or `Object.keys(l)`

In the regards of performance I don't know which one should be preferred.
Or maybe there is other ways to do this.

**Comment by benzen** [comment]
Fooling arround this is due to the behavior of Array.concat

**Comment by CrossEye** [comment]
We've always insisted that Ramda's default iterable type is a *list*.  Because the language does not have a built-in pair-based list structure (no CONS/CAR/CDR) we adopt dense arrays as the closest analog available.  But if you supply a sparse array to a function looking for a list, we let the chips fall where they may.  We don't even try to compensate.

**Comment by benzen** [comment]
I m ok with that décision. Still i think that the doc should warn about this on all func that use Array.concat

Benjamin Dreux

**Comment by buzzdecafe** [comment]
> the doc should warn

agreed

**Comment by CrossEye** [comment]
Yes, perhaps a link to a common, "Handling of sparse arrays" note.
